server/src/main/java/org/elasticsearch/common/compress/lz4/LZ4SafeUtils.java

```diff
@@ -0,0 +1,257 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.elasticsearch.common.compress.lz4;
+
+import net.jpountz.lz4.LZ4Exception;
+import net.jpountz.util.Utils;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+import java.util.Arrays;
+
+import static org.elasticsearch.common.compress.lz4.LZ4Constants.LAST_LITERALS;
+import static org.elasticsearch.common.compress.lz4.LZ4Constants.ML_BITS;
+import static org.elasticsearch.common.compress.lz4.LZ4Constants.ML_MASK;
+import static org.elasticsearch.common.compress.lz4.LZ4Constants.RUN_MASK;
+
+/**
+ * This file is forked from https://github.com/lz4/lz4-java. In particular, it forks the following file
+ * net.jpountz.lz4.LZ4SafeUtils.
+ *
+ * It modifies the original implementation to use Java9 array mismatch method and varhandle performance
+ * improvements. Comments are included to mark the changes.
+ */
+enum LZ4SafeUtils {
+    ;
+
+    // Added VarHandle
+    private static final VarHandle intPlatformNative = MethodHandles.byteArrayViewVarHandle(int[].class, Utils.NATIVE_BYTE_ORDER);
+    private static final VarHandle longPlatformNative = MethodHandles.byteArrayViewVarHandle(long[].class, Utils.NATIVE_BYTE_ORDER);
+
+    static int hash(byte[] buf, int i) {
+        return LZ4Utils.hash(SafeUtils.readInt(buf, i));
+    }
+
+    static int hash64k(byte[] buf, int i) {
+        return LZ4Utils.hash64k(SafeUtils.readInt(buf, i));
+    }
+
+    static boolean readIntEquals(byte[] buf, int i, int j) {
+        return SafeUtils.readInt(buf, i) == SafeUtils.readInt(buf, j);
+    }
+
+    static void safeIncrementalCopy(byte[] dest, int matchOff, int dOff, int matchLen) {
+        for (int i = 0; i < matchLen; ++i) {
+            dest[dOff + i] = dest[matchOff + i];
+        }
+    }
+
+    // Modified wildIncrementalCopy to mirror version in LZ4UnsafeUtils
+    static void wildIncrementalCopy(byte[] dest, int matchOff, int dOff, int matchCopyEnd) {
+        if (dOff - matchOff >= 4) {
+            if (dOff - matchOff < 8) {
+                copy8Bytes(dest, matchOff, dest, dOff);
+                dOff += dOff - matchOff;
+            }
+        } else {
+            for(int i = 0; i < 4; ++i) {
+                dest[dOff + i] = dest[matchOff + i];
+            }
+
+            dOff += 4;
+            matchOff += 4;
+            int dec = 0;
+
+            assert dOff >= matchOff && dOff - matchOff < 8;
```

Comment:
I do notice that this is a copy from original, but I wonder how `dOff - matchOff` can ever be >= 4 since we handle that case above (and then add 4 to both here)? I feel like I am missing something obvious, help me :slightly_smiling_face:

Reply:
No idea. I accidentally had copied the class file opposed to the original source. I modified this PR to copy the original source which logically makes more sense. If it is reorder in the class file, well 🤷‍♂️.